taxi/taxi.py

```diff
@@ -197,23 +204,59 @@ def step(self, action):
                     if (self.agent.position == p.position).all():
                         p.intaxi = True
                         self.passenger = p
+                        pickup = True
                         break  # max one passenger per taxi
+		
+                if not pickup:
+                    bad_pickup = True
             else:
                 # drop off?
                 dropoff_clear = True
                 for p in (p for p in self.passengers if p is not self.passenger):
                     if (p.position == self.passenger.position).all():
                         dropoff_clear = False
+                        bad_dropoff = True
                         break
+
                 if dropoff_clear:
                     self.passenger.intaxi = False
+                    
+		    #drop of current passenger at right depot
+                    depot_position = self.depots[self.passenger.goal].position
+
+                    if np.all(depot_position==self.passenger.position):
+                        good_dropoff = True
+                    else:
+                        bad_dropoff = True
+
                     self.passenger = None
+		    
+		    
+
         s = self.get_state()
-        if (self.goal is not None) or not self.check_goal(s):
+        if (self.goal is None) or not self.check_goal(s):
             done = False
         else:
             done = True
-        r = -1.0 if not done else 1000
+
+
+        #r = 0.0 if not done else 1.0
+        
+        if pickup:
+            r = +5.0
+        elif action==4 and (bad_dropoff or bad_pickup):
+           r = -10.0
+        elif action==4 and good_dropoff:
+            r = +20.0
+        else:
+            r = -1.0
```

Comment:
Wrap these different reward functions in an if/else block. The default should be the 0/1 loss, and the taxi constructor can have a new flag `use_gym_rewards=False` that allows for overriding the default reward function.